### PR TITLE
Implement TypeScript assertions in http-assert types

### PR DESCRIPTION
**Pull request by remcohaszing**
Basically this means one could write:

```ts
import { Context } from 'koa';

function myController(ctx: Context) {
  const { answer } = ctx.request.body;
  ctx.assert(answer === 42, 'Wrong')
  // The type of answer is 42 now.
  let correctAnswer: 42 = answer;
}
```

Because the `asserts` keyword was introduced in TypeScript 3.7, the old types were moved into the `ts3.6` folder. End of support for TypeScript 3.6 is this month, so I’m not sure if I should just remove them.

The try / catch clause was removed from tests, because it wasn’t used for testing the `http-assert` types.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: N/A
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/http-assert/http-assert-tests.ts [hunk 1]
```diff
@@ -1,26 +1,115 @@
...
+
+(() => {
+    const value = unknown();
+    httpAssert(value === null, 400, "message", { foo: "bar" });
+
+    // $ExpectType null
+    value;
+})();
+
+// assert.ok()
```
Comment (pkuczynski) [inline types/http-assert/http-assert-tests.ts:55]: I guess this is a leftover?

File: types/http-assert/http-assert-tests.ts [hunk 2]
```diff
@@ -1,26 +1,115 @@
...
+
+    // $ExpectType null
+    value;
+})();
+
+// Other assertions
 
 const status = 401;
-const message = 'some error message';
+const message = "some error message";
```
Comment (pkuczynski) [inline types/http-assert/http-assert-tests.ts:100]: Can you please revert changing `'` to `"` so the diff is more sensible?

File: types/http-assert/index.d.ts [hunk 3]
```diff
@@ -11,30 +11,30 @@
  * @param msg the message of the error, defaulting to node's text for that status code
  * @param opts custom properties to attach to the error object
  */
-declare function assert(value: any, status?: number, msg?: string, opts?: {}): void;
-declare function assert(value: any, status?: number, opts?: {}): void;
+declare function assert(value: any, status?: number, msg?: string, opts?: Record<string, any>): asserts value;
```
Comment (pkuczynski) [inline types/http-assert/index.d.ts:14]: I am just wondering if signature should not allow overriding `opts` type:

```
declare function assert<T = Record<string, unknown>>(value: any, status?: number, msg?: string, opts?: T): asserts value;
```

File: types/http-assert/index.d.ts [hunk 4]
```diff
@@ -11,30 +11,30 @@
...
+declare function assert(value: any, status?: number, opts?: Record<string, any>): asserts value;
 
 declare namespace assert {
     /**
      * @param status the status code
      * @param msg the message of the error, defaulting to node's text for that status code
      * @param opts custom properties to attach to the error object
      */
-    type Assert = <T>(a: T, b: T, status?: number, msg?: string, opts?: {}) => void;
+    type Assert = <T>(a: T, b: T, status?: number, msg?: string, opts?: Record<string, any>) => void;
```
Comment (pkuczynski) [inline types/http-assert/index.d.ts:23]: Shouldn't return type here and few lines below not be changed to `asserts a`?

File: types/koa/test/default.ts [hunk 5]
```diff
@@ -33,7 +33,7 @@ app.use<{ a: boolean }>(async (ctx, next) => {
     await next();
 });
 
-app.use((ctx, next) => {
+app.use((ctx: Koa.Context, next) => {
```
Comment (pkuczynski) [inline types/koa/test/default.ts:36]: This change is not related to the subject of this PR and should be done in a separate PR.

File: types/koa/test/index.ts [hunk 6]
```diff
@@ -37,7 +37,7 @@ app.use<{}, UserContext>(async ctx => {
     ctx.user = {};
 });
 
-app.use((ctx, next) => {
+app.use((ctx: Koa.Context, next) => {
```
Comment (pkuczynski) [inline types/koa/test/index.ts:40]: This change is not related to the subject of this PR and should be done in a separate PR.

**Review by pkuczynski — CHANGES_REQUESTED** [review]
Awesome work! Thank you! Left few inline comments...

File: types/http-assert/http-assert-tests.ts (hunk 2 above)
Comment (remcohaszing) [reply types/http-assert/http-assert-tests.ts:100]: I reformatted the file to use single quotes, as per the Prettier configuration as this was basically the only line that stayed the same.

The original file was moved into the ts3.6 folder and a new file with the same name was created where coincidentally some lines are the same.

File: types/http-assert/index.d.ts (hunk 3 above)
Comment (remcohaszing) [reply types/http-assert/index.d.ts:14]: There is no point in using a generic for one input parameter. This gives a slight performance loss when type checking and is disallowed by tslint ([no-unnecessary-generics](https://github.com/Microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md)).

File: types/http-assert/index.d.ts (hunk 4 above)
Comment (remcohaszing) [reply types/http-assert/index.d.ts:23]: This tests shallow equality between `a` and `b`. `asserts a` means it would test if `a` is truthy.

This type may be enhanced somehow, but I’m not sure how at this point.

File: types/koa/test/default.ts (hunk 5 above)
Comment (remcohaszing) [reply types/koa/test/default.ts:36]: I understand why you think so, but it really is.

Due to a bug in TypeScript the type of an assertion function or object containing an assertion function must be explicitly types. It doesn’t work if the type is inferred from `Koa.Middleware` or `new Koa().use()`.

```ts
app.use((ctx) => {
  // The result is used, so this is fine
  const a = ctx.throw();

  // This is not an assertion, so this is fine
  ctx.toString();

  // This only works if ctx is explicitly typed
  // Assertions require every name in the call target to be declared with an explicit type annotation. ts(2775)
  ctx.throw();
});
```

Related bugs: [microsoft/TypeScript#34596](https://github.com/microsoft/TypeScript/issues/34596) [microsoft/TypeScript#36931](https://github.com/microsoft/TypeScript/issues/36931)

This is why I initially gave up on #52449

File: types/koa/test/index.ts (hunk 6 above)
Comment (remcohaszing) [reply types/koa/test/index.ts:40]: Same as above.

File: types/http-assert/http-assert-tests.ts [hunk 7]
```diff
@@ -1,26 +1,113 @@
 import httpAssert = require('http-assert');
-import { HttpError } from 'http-errors';
+
+declare function unknown(): unknown;
+
+// assert()
```
Comment (pkuczynski) [inline types/http-assert/http-assert-tests.ts:5]: Is this a leftover comment? Maybe it's a description of the test set, not sure. Maybe it's worth splitting the tests into two separate files as we did in `koa` package? This file becomes pretty long now...

File: types/http-assert/index.d.ts (hunk 3 above)
Comment (pkuczynski) [reply types/http-assert/index.d.ts:14]: Fair enough

File: types/http-assert/index.d.ts (hunk 4 above)
Comment (pkuczynski) [reply types/http-assert/index.d.ts:23]: You are right. This is not solvable at the moment or at least I don't know how...

File: types/koa/test/default.ts (hunk 5 above)
Comment (pkuczynski) [reply types/koa/test/default.ts:36]: Got it! Thanks for in-depth explanation...

**Review by pkuczynski — APPROVED** [review]

**Review by pkuczynski — APPROVED** [review]

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by pkuczynski** [comment]
Ready to merge

**Comment by remcohaszing** [comment]
Ready to merge
